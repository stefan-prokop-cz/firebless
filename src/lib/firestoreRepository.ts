import CollectionReference = FirebaseFirestore.CollectionReference;
import SetOptions = FirebaseFirestore.SetOptions;
import Query = FirebaseFirestore.Query;
import Precondition = FirebaseFirestore.Precondition;
import {
    bulkCreate,
    bulkDelete,
    bulkUpdate,
    create,
    detail,
    detailById,
    getModelData,
    list,
    Pagination,
    remove,
    removeById,
    update,
    updateById,
    upsert,
} from './queries';

export const bind = <Model extends { id?: string }>(collection: CollectionReference) => {
    return {
        bulkCreate: (data?: Partial<Model>[]) => bulkCreate<Model>(collection, data),
        bulkUpdate: (data?: Partial<Model>[], options?: SetOptions) => bulkUpdate<Model>(collection, data, options),
        bulkDelete: (ids?: string[], precondition?: Precondition) => bulkDelete(collection, ids, precondition),
        create: async (data?: Partial<Model>) => {
            const doc = await create<Model>(collection, data);
            return getModelData(doc) as Model;
        },
        delete: (params?: Partial<Model>, options?: Query) => remove<Model>(collection, params, options),
        deleteById: (id: string, precondition?: Precondition) => removeById(collection, id, precondition),
        detail: async (params?: Partial<Model>, options?: Query) => {
            const doc = await detail<Model>(collection, params, options);
            return getModelData<Model>(doc);
        },
        detailById: async (id: string) => {
            const doc = await detailById<Model>(collection, id);
            return getModelData<Model>(doc);
        },
        getCollection: () => collection,
        getFirestore: () => collection.firestore,
        list: async (filters?: Partial<Model> & Pagination, options?: Query) => {
            const docs = await list<Model>(collection, filters, options);
            return docs.map(doc => getModelData<Model>(doc)) as Model[];
        },
        update: (params?: Partial<Model>, data?: Partial<Model>, options?: SetOptions) => update<Model>(collection, params, data, options),
        updateById: async (id: string, data?: Partial<Model>, options?: SetOptions) => {
            const doc = await updateById<Model>(collection, id, data, options);
            return getModelData(doc) as Model;
        },
        upsert: async (data?: Partial<Model>, options?: SetOptions) => {
            const doc = await upsert<Model>(collection, data, options);
            return getModelData(doc) as Model;
        },
    };
};
