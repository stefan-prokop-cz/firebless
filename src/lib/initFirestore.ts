import { Firestore, Settings } from '@google-cloud/firestore';

export default (config: Settings) => new Firestore(config);
