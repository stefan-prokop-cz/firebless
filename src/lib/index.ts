import * as defaultFirestoreRepository from './firestoreRepository';
export { default as initFirestore } from './initFirestore';
export {
    applyFilters,
    applyPagination,
    bulkCreate,
    bulkDelete,
    bulkUpdate,
    create,
    detail,
    detailById,
    getDocument,
    getModelData,
    list,
    Pagination,
    remove,
    removeById,
    update,
    updateById,
    upsert,
} from './queries';
export { defaultFirestoreRepository };
