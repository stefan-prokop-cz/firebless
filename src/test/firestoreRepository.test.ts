import { Firestore } from '@google-cloud/firestore';
// @ts-ignore
import MockFirebase from 'mock-cloud-firestore';
import { defaultFirestoreRepository, initFirestore } from '../lib/index';
import { idMatcher } from './matchers';

const fixtures = {
    __collection__: {
        users: {
            __doc__: {
                user_a: {
                    age: 15,
                    name: 'user',
                    surname: 'a',
                    birthDate: new Date('2019-01-01'),
                },
                user_b: {
                    age: 30,
                    name: 'user',
                    surname: 'b',
                    birthDate: new Date('2018-01-01'),
                },
            },
        },
    },
};

const firebase = new MockFirebase(fixtures);
const firestore = firebase.firestore();

interface User {
    id?: string;
    name: string;
    surname: string;
    age: number;
    birthDate: Date;
}

const userRepository = defaultFirestoreRepository.bind<User>(firestore.collection('users'));

describe('Firestore repository', () => {
    it('Initialization', () => {
        const db = initFirestore({});
        expect(db).toBeInstanceOf(Firestore);
    });
    it('List', async () => {
        const users = await userRepository.list();
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(2);
        expect(users[0]).toMatchSnapshot();
    });
    it('List with filters', async () => {
        const users = await userRepository.list({ age: 15 });
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(1);
        expect(users[0]).toMatchSnapshot();
    });
    it('List with options', async () => {
        const users = await userRepository.list({}, userRepository.getCollection().where('age', '>=', 16));
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(1);
        const { age } = users[0];
        expect(age).toBe(30);
    });
    it('Empty list', async () => {
        const users = await userRepository.list({ age: 20 },
            userRepository.getCollection().where('surname', '==', 'a'));
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(0);
    });
    it('List with ordering', async () => {
        const users = await userRepository.list({}, userRepository.getCollection().orderBy('age', 'desc'));
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(2);
        const u1 = users[0];
        const u2 = users[1];
        expect(u1.age).toBeGreaterThan(u2.age);
    });
    it('List with filters & ordering', async () => {
        const users = await userRepository.list({ name: 'user' }, userRepository.getCollection().orderBy('age', 'asc'));
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(2);
        const u1 = users[0];
        const u2 = users[1];
        expect(u1.age).toBeLessThan(u2.age);
    });
    it('Detail with filters & ordering', async () => {
        await userRepository.create({ age: 100, id: 'user_l', surname: 'l', name: 'user' });
        const first = await userRepository.detail({ name: 'user' }, userRepository.getCollection().orderBy('age', 'desc'));
        const second = await userRepository.detail({ name: 'user' }, userRepository.getCollection().orderBy('age', 'asc'));
        expect(first?.age).toBeGreaterThan(second!.age);
    });
    it('Create', async () => {
        const user: User = {
            age: 10,
            birthDate: new Date('2020-01-01'),
            name: 'Test',
            surname: 'Testovic',
        };
        const createdUser = await userRepository.create(user);
        expect(createdUser).toMatchSnapshot(idMatcher);
        const users = await userRepository.list();
        expect(Array.isArray(users)).toBe(true);
        expect(users).toHaveLength(4);
    });
    it('Update by ID', async () => {
        const user = await userRepository.detail({ surname: 'a' });
        expect(!user).toBe(false);
        const updatedUser = await userRepository.updateById(user!.id!, { surname: 'c' });
        expect(updatedUser.surname).toBe('c');
        const u = await userRepository.detail({ surname: 'a' });
        expect(u).toBeNull();
        expect(updatedUser).toMatchSnapshot();
    });
    it('Upsert', async () => {
        const data: User = {
            name: 'Stefan',
            surname: 'Prokop',
            age: 1,
            birthDate: new Date('2020-01-01'),
            id: '123',
        };
        const user = await userRepository.upsert(data);
        const updatedUser = await userRepository.upsert({ ...data, surname: 'Stefan' });
        expect(user.surname === updatedUser.surname).toBe(false);
        expect(user.id === updatedUser.id).toBe(true);
        expect(updatedUser).toMatchSnapshot();
    });
    it('Detail by ID', async () => {
        const user = await userRepository.detailById('123');
        expect(user).toMatchSnapshot();
    });
    it('Empty detail', async () => {
        const user = await userRepository.detailById('abcdefg');
        expect(user).toBeNull();
    });
    it('Detail with filters', async () => {
        const user = await userRepository.detail({ name: 'user' });
        expect(user).toMatchSnapshot();
    });
    it('Delete by ID', async () => {
        try {
            await userRepository.deleteById('123');
        } catch (_) {}
        const user = await userRepository.detailById('123');
        expect(user).toBeNull();
    });
    it.skip('Bulk create', async () => {
        const prevUsers = await userRepository.list();
        await userRepository.bulkCreate([
            { age: 5, name: 'Philip', surname: 'Palencia' },
            { age: 11, name: 'Dwayne', surname: 'Carter' },
        ]);
        const users = await userRepository.list();
        expect(Array.isArray(prevUsers)).toBe(true);
        expect(Array.isArray(users)).toBe(true);
        expect(prevUsers.length).toBeLessThan(users.length);
        expect(users.length - prevUsers.length).toBe(2);
    });
    it('Bulk update', async () => {
        const prevUsers = await userRepository.list();
        await userRepository.bulkUpdate(prevUsers.map(u => ({ id: u.id, name: 'Bulk', surname: 'Update' })));
        const users = await userRepository.list();
        expect(Array.isArray(prevUsers)).toBe(true);
        expect(Array.isArray(users)).toBe(true);
        expect(prevUsers.length).toEqual(prevUsers.length);
        users.forEach(user => {
            expect(user.name).toEqual('Bulk');
            expect(user.surname).toEqual('Update');
        });
    });
    it('Bulk delete', async () => {
        const prevUsers = await userRepository.list();
        await userRepository.bulkDelete(prevUsers.filter(u => u.id).map(u => u.id) as string[]);
        const users = await userRepository.list();
        expect(Array.isArray(prevUsers)).toBe(true);
        expect(Array.isArray(users)).toBe(true);
        expect(prevUsers.length).toBeGreaterThan(users.length);
        expect(users.length).toEqual(0);
    });
});
