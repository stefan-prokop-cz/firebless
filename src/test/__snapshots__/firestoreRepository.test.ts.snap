// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Firestore repository Create 1`] = `
Object {
  "age": 10,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": Any<String>,
  "name": "Test",
  "surname": "Testovic",
  "updatedAt": undefined,
}
`;

exports[`Firestore repository Detail by ID 1`] = `
Object {
  "age": 1,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "123",
  "name": "Stefan",
  "surname": "Stefan",
  "updatedAt": undefined,
}
`;

exports[`Firestore repository Detail with filters 1`] = `
Object {
  "age": 15,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "user_a",
  "name": "user",
  "surname": "c",
  "updatedAt": undefined,
}
`;

exports[`Firestore repository List 1`] = `
Object {
  "age": 15,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "user_a",
  "name": "user",
  "surname": "a",
  "updatedAt": undefined,
}
`;

exports[`Firestore repository List with filters 1`] = `
Object {
  "age": 15,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "user_a",
  "name": "user",
  "surname": "a",
  "updatedAt": undefined,
}
`;

exports[`Firestore repository Update by ID 1`] = `
Object {
  "age": 15,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "user_a",
  "name": "user",
  "surname": "c",
  "updatedAt": undefined,
}
`;

exports[`Firestore repository Upsert 1`] = `
Object {
  "age": 1,
  "birthDate": Object {
    "toDate": [Function],
  },
  "createdAt": undefined,
  "id": "123",
  "name": "Stefan",
  "surname": "Stefan",
  "updatedAt": undefined,
}
`;
